### DISCUSSION TITLE
タグの名寄せ方針を考えています。ご意見ください 🙏

### DISCUSSION THREAD

**Discussion by kyntk** [increments/qiita-discussions#205] [要望・アイデア・不具合]
こちらのDiscussionで、タグの名寄せのリクエストをいただいています。
https://github.com/increments/qiita-discussions/discussions/80#discussioncomment-2474785

現在、Qiitaはタグの名寄せの明確な方針がないため、毎回どのタグをどのタグに名寄せをするのかの判断に悩んでおります。
タグについては、運営が意図的に寄せることをあまりしたくないので、方針を決めてそれに沿っていきたいと考えています。

今回、名寄せの方針を考えてみましたので、確定前に皆様にご意見をいただきたいです 🙏 

## 前提: タグの名寄せとは

`C` , `C言語` のようにタグが複数に分かれているものを `C` に統一して記事に紐付け直すことを名寄せと呼んでいます。

名寄せをすることで、タグで調べた時に見つけやすく、書き手としても見つけてもらいやすくなります。

## タグの名寄せ方針

Qiitaではタグの名称を自由につけることができるので、基本的に運営主導で名寄せをすることはあまりしません。
ただし、以下のような時に名寄せをすることがあります。

- 明らかな誤字のとき (`Rails`/`Ralis`)
- caseの違いだけのとき (`common-lisp`/`CommonLisp`)
- 英語と、読みをカタカナにしているだけとき (`network`/`ネットワーク`)
- ~~`C#`等の記号を含むとき~~
    - ([こちらで議論した経緯](https://github.com/increments/qiita-discussions/discussions/205#discussioncomment-2583854)により削除)
- 正式名称の前後に余計な記号や文字がついている（先頭の`#`や、末尾の`,`など）

これらの時に、今までは概ね記事数の多い方に寄せていたのですが、判断が難しいものも出てきたので、今後は以下のようにしていこうと考えています。

- 明らかな誤字のとき (`Rails`/`Ralis`)
    - → **誤字をしていない**タグへの名寄せ
- caseの違いだけのとき (`common-lisp`/`CommonLisp`)
    - → **公式サイト**を見て正しそうなもの。ただし、 `Common Lisp`のようにスペースがあるものは**スペースを削るだけ**の方を優先
- 英語と、読みをカタカナにしているだけの英語のとき (`network`/`ネットワーク`)
    - → 片方があまり存在せず、記事数に**大差がついているときは多い方**へ揃え、どちらも多い場合は **英語** に揃える
- 正式名称の前後に余計な記号や文字がついている（先頭の`#`や、末尾の`,`など）
    - → **正式名称に寄せる**

これは既存のタグを自動的に名寄せするというよりも、判断の軸として使う想定で、今の所既存のタグを自動ですべて名寄せする予定はありません。

## ご意見をください 🙇 

皆様のご意見をもとに方針を決めていきたいと思います。
もしタグを決める時に、こう考えているなどもありましたらご意見ください！🙇

**Comment by shiracamus** [comment]
> 先頭に#や、末尾に,がつくとき
> → 削除する

これをしていただけるだけで、私が2013年からコツコツ続けてきた編集リクエストをしなくて済むので大変ありがたいです。
全角文字で付ける人がいたり、先頭にカンマを付けてしまう人もいるので、正規表現で処理するのがいいかと思います。
他には、タグを `[ ]` で囲んだり、`@` で始める人がいたりするけど、それは気付いたときに編集リクエストすればいいかと思います。
それと、`#`だけ、`,`だけ のタグを指定されることもあるので予期せぬ動作をしないようご注意ください。

先頭： `^[#＃,，、]`
末尾： `[,，、]$`

**Reply by kskwtnk** [reply]
@shiracamus
ありがとうございます！

現状`#`だけ、`,`だけになっているものの処理は考えます！

**Reply by shiracamus** [reply]
現状のタグ一覧で、該当するタグがどれくらいあるか確認してみてはいかがでしょうか？
タグ一覧は全部で749ページ。
https://qiita.com/tags?page=749
スクレイピングやAPIで取り出すのはサーバに負荷をかけてしまうので、サーバ内でサクッと一覧取得して調査していただくのがいいかと思います。

**Reply by shiracamus** [reply]
> これは既存のタグを自動的に名寄せするというよりも、判断の軸として使う想定で、今の所既存のタグを自動ですべて名寄せする予定はありません。

前後の余計な記号削除は自動化して欲しいです。日々余計な記号が付いた新規タグが生まれている状況なので、手動で名寄せ設定追加するのは非現実的です。
自動削除後に、正式名称への名寄せで「#で始まるタグ」や「カンマで終わるタグ」に再び変更される仕組みであれば、不満を言う人はいないと思います。誰かが「 #hogehoge を正式名称にして名寄せ設定してくれ」といったリクエストを出す必要はありますけど。

**Reply by kyntk** [reply]
@shiracamus ありがとうございます。

1つ1つ対応するのは現実的ではないというのはおっしゃる通りなので、方法を検討します。
対応方法として、強制的に削除する以外にも、サジェストを出すなど、いくつか方法が考えられるので、引き続き検討していきます :pray:

**Reply by boc1211** [reply]
「#後で直す」みたいなものだと先頭のシャープが余計なものなのか意図的なものなのか機械的に判断しづらいのかな、と横から見ていて思いました。
正式名称があるようなものでも無いですし。
これまで見てきてそんなケースは殆ど無かったんだとは思いますが、自動化するとなるとそういうことも考慮しないといけないのかな、と。

**Reply by kyntk** [reply]
@boc1211 おっしゃる通りです...

なので、完全に自動化するのは難しいかもしれませんが、誤ったタグが増えてしまうのは防ぎたいので、いい落としどころを検討していきます！

**Reply by shiracamus** [reply]
twitterのハッシュタグは全部 #付き ですが「##後で消す」なんて#を2個付けたりしないですよね？
Qiitaのタグはフォローできてタイムラインで新着記事を読めるので、同じ興味・関心を持つ人や記事を集めるためのキーワードとして私は使っています。動作プラットフォームや使用ライブラリ名を含めたりもしますけど。
タグは「この指とまれ」的な存在で、名寄せして多くのフォロワーに記事が届くようになって欲しいです。

**Reply by k1350** [reply]
@shiracamus 自動化してほしいという話は

https://github.com/increments/qiita-discussions/discussions/67

に別途立てられていたようですから、そちらで議論されてはどうですか？  
なぜかクローズのタグついてますが……。

ここは名寄せ方針の discussion ですので、違う話題を混ぜないほうが良いと思います。

**Comment by munierujp** [comment]
おおむねいいと思いますが、表記が揺れがちなJS関連のタグについてはどうでしょうか？

[Qiitaの◯◯.jsタグの表記ゆれと最も使われているタグ一覧 - Qiita ](https://qiita.com/potato4d/items/23023664dbfdbbd77b98)

なんとなく著名なライブラリ・フレームワークは正式名称に名寄せされていそうですが、たとえば[nuxt](https://qiita.com/tags/nuxt)（191記事）と[nuxt.js](https://qiita.com/tags/nuxt.js)（3636記事）など、名寄せされていないものもあるようです。
運営側で名寄せできるのであれば、各タグの記事の多さとは関係なく正式名称に寄せるというのは良いアプローチだと思います。

**Reply by kskwtnk** [reply]
@munierujp 
ありがとうございます！

正式名称に寄せていくつもりですが、いきなりだと驚く人もいるかもしれないので周知をしっかりしてからの実行をしよう、と考えています！

**Comment by radian-jp** [comment]
Stack Overflowでは、一定以上の権限を持つユーザーがタグの別名を提案出来るようなシステムがありますが、
このへん参考になりませんでしょうか。
[タグの別名](https://ja.stackoverflow.com/tags/synonyms)
[タグの別名の作成](https://ja.stackoverflow.com/help/privileges/suggest-tag-synonyms)

**Reply by kskwtnk** [reply]
@radian-jp 
ありがとうございます！

こちらのようなシステム、良いですね！ありがとうございます！
今回はまず一旦顕在化しているものを名寄せする対応に留めようと思っていますが、ゆくゆくはこういった仕組みにして生きたいです！

**Comment by k1350** [comment]
> - C#等の記号を含むとき
>     - → csharpのように置き換える
> - 先頭に#や、末尾に,がつくとき
>     - → 削除する

これを判断の軸とするのには反対です。

判断の軸として最も優先すべきなのは「正式名称に寄せる」であるべきだと思います。
私の認識では少なくとも C# は正式名称です。
私は C# について検索するとき初手で csharp とは打ち込みません。
正式名称に記号が入っており、Qiita のタグにその記号を使うことが可能ならば、正式名称のほうに寄せていただきたいです。

そもそも「C# を csharp に置換する」という話は「誤字」「caseの違い」などとは別問題ではありませんか？
たぶん「URL 化するときに困る」みたいな Qiita のシステム都合であり、

> 名寄せをすることで、タグで調べた時に見つけやすく、書き手としても見つけてもらいやすくなります。

というユーザーにとっての利便性とは無関係だと思うのですが……。（C# というタグのページは /tags/csharp となっているし、C++ は /tags/cpp となっているので）

仮に正式名称の一部であっても記号を入力されるとシステム的に困るのであれば、そもそも記号を入力できないようにしてほしいです。
書き手としては、入力できる以上は正式名称でタグをつけようと思いますので……。

**Reply by kyntk** [reply]
@k1350 ご意見いただきありがとうございます！

タグの名称とURL表記の話を混ぜてしまい申し訳ございません。

> C# というタグのページは /tags/csharp となっているし、C++ は /tags/cpp となっている

ご指摘いただいたようにC#、C++ではこのようになっており、この件は「タグの名寄せ」とはまた別の内容なので、タグの名寄せのルールとしては除外しようと思います。
C#のように、タグを付与したり検索する分には`C#`で使える状態を維持いたします。

**Reply by k1350** [reply]
@kyntk C# についての話のみ拾っていただいていますが

> 先頭に#や、末尾に,がつくとき
> → 削除する

これにも反対です。

理由は「正式名称に寄せる」という判断の軸が示されていないからです。  
C# の事例さえ書いていなければ『「正式名称に寄せる」を暗黙の前提としている話なのだな』と行間を読めたのですが、C# の事例が書いてあったのでそう都合よく解釈はできません。

`#` で始まる名称のプログラミング言語が今後登場しないとは言えませんので、`#` や `,` を削除するという方針には反対です。 
 
また他の方が仰るように、実際には `[]` で囲まれていたり `@` で始まったりするタグもあるそうですので、`#` や `,` を特別に考慮する必要性を感じません。  
「正式名称の前後に余計な記号や文字がくっついている」ケースは全部まとめて「正式名称に寄せる」という方針で片付けられると思います。

私の意見を再度書きますが、判断の軸として最も優先すべきなのは「正式名称に寄せる」であるべきだと思います。

**Reply by radian-jp** [reply]
> #で始まる名称のプログラミング言語が今後登場しないとは言えませんので、#や , を削除するという方針には反対です。

この辺りは、そういったものが登場してから対応でよいのではという気がしないでもないですが。
先頭"#"はTwitterのハッシュタグと間違えて入れてる人がそこそこ居る印象なので、除去したいんじゃないですかね。
正式名称に寄せるというのは同意します。投稿の時点で、タグのtypoや別名を検出して正式名称に変換してくれると、名寄せの手間も減るでしょうし一番楽そうではあります😄

**Reply by kyntk** [reply]
@k1350 余計な記号の件についても申し訳ございません。 :bow:

方針の中に「先頭に#や、末尾に,」を取り上げたのは、誤って付けられていることが多いためで、意図していることは @k1350 のおっしゃるとおりです！
いただいた言葉を使わせていただき、Disucussionの記述も訂正いたしました！:bow:

ご意見ありがとうございます！

**Reply by kyntk** [reply]
@radian-jp ありがとうございます！

> 間違えて入れてる人がそこそこ居る

おっしゃるとおりでした！
ただ、正確さに欠ける部分もあるため上記にあるように訂正いたしました！

> 投稿の時点で、タグのtypoや別名を検出して正式名称に変換してくれると、名寄せの手間も減るでしょうし一番楽そうではあります😄

こちらも今後検討していきます！

**Comment by kyntk** [comment]
皆様、たくさんのご意見をいただきありがとうございます :bow: 

皆様の意見を踏まえまして、タグの名寄せは以下の方針でいきたいと思います！
また追ってBlogにてアナウンスいたします！

- 明らかな誤字のとき (`Rails`/`Ralis`)
    - → **誤字をしていない**タグへの名寄せ
- caseの違いだけのとき (`common-lisp`/`CommonLisp`)
    - → **公式サイト**を見て正しそうなもの。ただし、 `Common Lisp`のようにスペースがあるものは**スペースを削るだけ**の方を優先
- 英語と、読みをカタカナにしているだけの英語のとき (`network`/`ネットワーク`)
    - → 片方があまり存在せず、記事数に**大差がついているときは多い方**へ揃え、どちらも多い場合は **英語** に揃える
- 正式名称の前後に余計な記号や文字がついている（先頭の`#`や、末尾の`,`など）
    - → **正式名称に寄せる**

ご意見いただきありがとうございました！

**Comment by NurseAngel** [comment]
名寄せとはちょっと違う話なのですが、
タグの大文字小文字が勝手に変更されてしまうことがあるので少々困っています。
たとえば`APCu`が勝手に`ApCu`になる、など。

おそらく最初に投稿されたタグに合わせられるのだと思います。

[検索](https://qiita.com/tags/apcu)では大文字小文字を気にせず引っかかるのも仕方ないですが、
大文字小文字で違う意味になってしまう単語もあるので、せめて[個別記事](https://qiita.com/rana_kualu/items/45afbb591eb8195d2518)上では投稿したそのままの表示にしてほしいです。
あるいは、せめて小文字に統一するとかであればまだマシかなと。

**Reply by tanaka-asami415** [reply]
@NurseAngel 

ご意見いただきありがとうございます！

>おそらく最初に投稿されたタグに合わせられるのだと思います。

ご指摘の通りです。一度登録された名称に対して
大文字小文字が区別されない設計になっております。

改修すると他の記事にも影響が大きいため、変更は難しいのですが、
タグの表示部分であれば、タグごとに修正することも可能です。
この場合、個別の記事ごとではなく、タグの設定変更になるので、
明らかな誤字と判断できる場合のみの対応としたいです。

`ApCu`に関しては、`APCu`へ修正させていただきます。

他にもありましたら、こちら [#80 ](https://github.com/increments/qiita-discussions/discussions/80)からリクエストください🙏

**Comment by yasokada** [comment]
記事を書くときに考えて付けたタグが別物になるのは、違和感ですね。

間違ったタグを入力した時に「もしかして」で正しいタグを推薦する程度はいいです。
ユーザが用意したデータが運営によって書き換えられることには違和感があります。

タグ検索のことを考えるとTag Synonymsというのは一つの解決策である気もしますが、実装・運用コストが高いでしょうかね。
https://stackoverflow.com/tags/synonyms

ユーザの中には「考えて」タグを決めている人もいることを考慮して、名寄せを実行する前には、ユーザ設定で「名寄せOFF」できるように考えてほしいところです。

**Reply by tanaka-asami415** [reply]
@yasokada 
ご意見ありがとうございます🙏

> 記事を書くときに考えて付けたタグが別物になるのは、違和感ですね。

私たちも、タグの名称を運営の独断で変えてしまうことは、
基本的にはあまりするべきではないと考えております。

ただし、明らかな誤字などの場合や、正式名称があるものに関しては
混在したままですと困られているユーザー様も多く、名寄せを行うことにいたしました🙇
https://blog.qiita.com/replacing-tag-rule/

既存のタグを自動ですべて名寄せする予定は今のところございませんが
今回いただいたご意見は、今後の改善に参考とさせていただきます🙏

**Reply by yasokada** [reply]
誤字の訂正(Ralis -> Rails)は賛成できますが、「ネットワーク」を「network」にするとか、「C言語」を「C]にするとかはどうかと思うところはあります。

タグというのは複数のサービスで利用可能なもので、複数サービスでタグを一貫性あるようにつけているユーザもいるでしょうね。
その時に、他のサービスでは「Ｃ言語」で検索できるものがQiitaだけは「Ｃ」にしないといけないのは記事を作る人の検索の負担になりそうです。
そういう用語変更がいくつもあると、その人自身の検索時の負担になります。

「ネットワーク」を「network」に変更するのも同様です。日本語名タグで統一している人もいる場合、その人がQiitaで検索時に英語表記を使わないといけないのは負担でしょうね。

自分はタグを基本的に英語にしますが、タグテーブル上の閲覧性の観点から日本語を使う場合もあります。

**Reply by shiracamus** [reply]
検索結果をタグで絞り込むには、`tag:` でタグ名を指定する必要がありますよね？
名寄せしないと `tag:C` と `tag:C言語` で異なる検索結果が表示されていたのが、タグを名寄せすることでどちらを指定しても同じ検索結果が表示されると期待しているのですが、違いますかね？
英語かカタカナかも、どちらを指定しても同じ検索結果が表示されるのであれば有難いと思うのですが。

追記：検索してみたところ結果が違っていましたが、タグはどちらも `C` で絞り込まれていますね。検索処理にもタグの名寄せ対応（どちらも検索する）が必要そうです。
https://qiita.com/search?q=ポインタ+tag%3AC (767件)
https://qiita.com/search?q=ポインタ+tag%3AC言語 (144件)

**Reply by yasokada** [reply]
> 英語かカタカナかも、どちらを指定しても同じ検索結果が表示されるのであれば有難いと思うのですが。

Qiitaの検索機能として上記が選択的に使用できるのは賛成ですが、そのために記事のタグの紐付けを改変するのは反対ですね。
（理由はすでに書いたとおりです）。

「名寄せ」の定義を「記事のタグそのものを改変」でなく、「検索時に複数タグを考慮して検索」という形が取れればと思います。
それには実装コストが当然かかってしまいます。

- tagのor検索
- ユーザが入力したたキーワードをもとに関連するタグをテーブルから取り出してor検索
- 「直接検索」なのか「名寄せ検索」なのかの選択を可能に

**Reply by shiracamus** [reply]
Qiitaの検索は、タイトルと本文が対象で、タグは `tag:` 指定で絞り込むためのものかと思っていたのですが、違うのですかね？
検索以外のタグの役割として、タグをフォローすることでタイムラインで新着記事を読むことができます。
興味関心があるタイムラインの新着記事タイトルを見て記事を読み、LGTMしたりストックすることでホームやトレンドに良記事が表示され、多くの人の目にも止まるようになります。
`Python` はフォローしていたけれど `#Python` をフォローしてなくて「こんな記事あったっけ？」とか「この質問に回答したかったなぁ」とか思うこともありました。
なので、タグの名寄せで興味関心のある記事をある程度まとめて表示してくれるのは私にとっては有難いです。検索から話題が脱線してすみません。

**Reply by yasokada** [reply]
@shiracamus さん

> Qiitaの検索は、タイトルと本文が対象で、タグは tag: 指定で絞り込むためのものかと思っていたのですが、違うのですかね？

現状はそうですね。「ポインタ tag:C」と検索時に内部的に「ポインタ tag:C tag:C言語」の結果を返すのはどうでしょう、という提案です（実装コストはかかりますが）。

閲覧者の観点からは名寄せ自体はありがたいというのはそうですねと思います。
作成者にとってはどうでしょう、というところです。

一例にすぎませんが、「C言語」を「C」にした場合の弊害。
タグを複数サービスで一貫性のあるものにしたい場合、「C」に統一するとしますね。
ScrapBoxというサービスでは「C」というタグで検索すると「部分一致検索」となり、C, cslint, cppbuilder, css. cppなどが全て検索候補にあがります。
「C言語」で見つかりやすかったものが「C」では使い物にならないことになります。

「C++Builder」では、とあるサービスで弊害が出ることもあるので「cppBuilder」で907記事書いているというのもあります。

そのような作成者の意図があるということはご理解いただきたいところです。

**Reply by tanaka-asami415** [reply]
@yasokada @shiracamus 

貴重なご意見をありがとうございます🙏

現状の検索でも、`C`で検索して`cslint`, `cppbuilder`, `css`, `cpp`がヒットすることはありません。
ただ、逆に`C言語`で検索して`C`がヒットしないのは、不便になってしまいますので今後改善していきます🙇
現状の開発状況的に、今回の改修ではまず[こちら](https://blog.qiita.com/replacing-tag-rule/)の範囲のみで進めていければと思います🙏

**Reply by yasokada** [reply]
意図が伝わらなかったようで残念です。

**Comment by tanaka-asami415** [comment]
たくさんのご意見ありがとうございました 🙏

本件、クローズさせていただきます！
なお、今回決定した方針については、こちらでご確認いただけます。
[タグの名寄せ方針を決定しました](https://blog.qiita.com/replacing-tag-rule/)
